### PR TITLE
[Ascend]  Ljf/modify for aclnn ops

### PR DESCRIPTION
**Pull request by liujingfeng4A069**
## Motivation and Context

* Modify for matmul_all_reduce and prompt_flash_attention.
* Modify for paged_attention and rotary_embedding_v2.
* Modify for customized_flash_attention(_backward) and customized_flash_attention_varlen(_backward).
## Checklist
**Before PR**:

- [x] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by POI-WX — APPROVED** [review]

File: impl/ascend/aclnn/adaptor.hpp [hunk 1]
```diff
@@ -192,13 +195,16 @@ decltype(auto) convertType(T&& param) {
         return diopiDtypeToAclDataType(std::forward<T>(param));
     } else if constexpr (IsBoolStdArray<U>::value) {
         return createAclBoolArrayFromVector<std::tuple_size_v<U>>(std::forward<T>(param));
+    } else if constexpr (std::is_same_v<U, const char*>) {
```
Comment (yangbofun) [inline impl/ascend/aclnn/adaptor.hpp:198]: 把这里删了，加强下else里面的static_assert的条件。

File: impl/ascend/aclnn/adaptor.hpp [hunk 2]
```diff
@@ -192,13 +195,16 @@ decltype(auto) convertType(T&& param) {
...
+        return std::forward<T>(param);
     } else {
         static_assert(!std::is_class_v<U> && !std::is_pointer_v<U>);
```
Comment (yangbofun) [inline impl/ascend/aclnn/adaptor.hpp:201]:
```suggestion
        static_assert(!std::is_class_v<U> && !std::is_class_v<std::remove_pointer<U>>);
```

File: impl/ascend/aclnn/adaptor.hpp [hunk 3]
```diff
@@ -192,13 +195,16 @@ decltype(auto) convertType(T&& param) {
...
         return std::forward<T>(param);
     }
 }
 
-template <class T, class U = std::remove_reference_t<T>, std::enable_if_t<!std::is_class_v<U> && !std::is_pointer_v<U>, int> = 0>
+template <class T, class U = std::remove_reference_t<T>,
```
Comment (yangbofun) [inline impl/ascend/aclnn/adaptor.hpp:206]: 改成
template<typename T>
void releaseConverted(......){}

File: impl/ascend/functions_ext/flash_attention.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+#include "../aclnn/acl_scalar.hpp"
+#include "../aclnn/adaptor.hpp"
+
+namespace impl {
+namespace ascend {
+
+namespace {
+
+const int64_t bitNumber = 128;
+const int64_t uInt8BitNumber = 8;
```
Comment (yangbofun) [inline impl/ascend/functions_ext/flash_attention.cpp:16]: 放到函数内用static定义下。

**Review by yangbofun — APPROVED** [review]

**Pull request merged**